src/libstd/fs.rs

```diff
@@ -1069,13 +1069,43 @@ pub fn remove_dir_all<P: AsRef<Path>>(path: P) -> io::Result<()> {
 }
 
 fn _remove_dir_all(path: &Path) -> io::Result<()> {
+    let filetype = try!(symlink_metadata(path)).file_type();
+    if filetype.is_symlink() {
+        if cfg!(windows) {
+            // On Windows symlinks to files and directories removed differently
+            // we should remove only directories here and have error on file
+            remove_dir(path)
+        } else {
+            // On unix symlinks are always files
+            remove_file(path)
+        }
+    } else {
+        remove_dir_all_recursive(path)
+    }
+}
+fn remove_dir_all_recursive(path: &Path) -> io::Result<()> {
     for child in try!(read_dir(path)) {
-        let child = try!(child).path();
-        let stat = try!(symlink_metadata(&*child));
-        if stat.is_dir() {
-            try!(remove_dir_all(&*child));
+        let child = try!(child);
+        let child_path = child.path();
+        let child_type = try!(child.file_type());
+        if child_type.is_dir() {
+            try!(remove_dir_all_recursive(&*child_path));
         } else {
-            try!(remove_file(&*child));
+            if cfg!(windows) {
+                if child_type.is_symlink() {
+                    let target_type = try!(metadata(&*child_path)).file_type();
```

Comment:
>I found that I could create a symlink to a file with `symlink_dir`,

That's because Windows doesn't check the target of symbolic links when you create them. You still created a directory symbolic link. It just happens to be pointing at an invalid thing.